### ISSUE TITLE
GPUComputePassEncoder.setBindGroup() index parameter description incorrect

### ISSUE THREAD

**Issue by randfur** [mdn/content#26627]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/GPUComputePassEncoder/setBindGroup

### What specific section or headline is this issue about?

https://developer.mozilla.org/en-US/docs/Web/API/GPUComputePassEncoder/setBindGroup#parameters

### What information was incorrect, unhelpful, or incomplete?

> The index to set the bind group at. This corresponds to the binding value of the relevant entry object in the descriptor of the originating [GPUDevice.createBindGroup()](https://developer.mozilla.org/en-US/docs/Web/API/GPUDevice/createBindGroup) call that created the bindGroup.

### What did you expect to see?

I think the index corresponds to the @group(n) value in the WGSL code, not the `binding` values passed to `createBindGroup()`.

### Do you have any supporting links, references, or citations?

I don't know for sure as I'm still learning but it's what makes sense to me and it's what ChatGPT seems to be telling me when I ask it for an example of using multiple `@group()`s.

GPT example code:
```wgsl
[[group(0), binding(0)]] var<uniform> u0 : U0;
[[group(1), binding(0)]] var<uniform> u1 : U1;
[[group(0), binding(1)]] var<uniform> u2 : U2;
[[group(1), binding(1)]] var<uniform> u3 : U3;
```

```js
// Create the uniform buffers
const u0 = device.createBuffer({
    size: /* size of U0 buffer */,
    usage: GPUBufferUsage.UNIFORM
});
const u1 = device.createBuffer({
    size: /* size of U1 buffer */,
    usage: GPUBufferUsage.UNIFORM
});
const u2 = device.createBuffer({
    size: /* size of U2 buffer */,
    usage: GPUBufferUsage.UNIFORM
});
const u3 = device.createBuffer({
    size: /* size of U3 buffer */,
    usage: GPUBufferUsage.UNIFORM
});

// Define the bind groups for the buffers
const bindGroup0 = device.createBindGroup({
    layout: pipeline.getBindGroupLayout(0),
    entries: [
        {
            binding: 0,
            resource: {
                buffer: u0
            }
        },
        {
            binding: 1,
            resource: {
                buffer: u2
            }
        }
    ]
});
const bindGroup1 = device.createBindGroup({
    layout: pipeline.getBindGroupLayout(1),
    entries: [
        {
            binding: 0,
            resource: {
                buffer: u1
            }
        },
        {
            binding: 1,
            resource: {
                buffer: u3
            }
        }
    ]
});

// Bind the buffers and dispatch the compute shader
const encoder = device.createCommandEncoder();
const pass = encoder.beginComputePass();
pass.setPipeline(computePipeline);
pass.setBindGroup(0, bindGroup0);
pass.setBindGroup(1, bindGroup1);
pass.dispatch(/* ... */);
pass.endPass();
device.queue.submit([encoder.finish()]);
```

### Do you have anything more you want to share?

The same will apply to https://developer.mozilla.org/en-US/docs/Web/API/GPURenderPassEncoder/setBindGroup.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/api/gpucomputepassencoder/setbindgroup`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/GPUComputePassEncoder/setBindGroup
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/api/gpucomputepassencoder/setbindgroup/index.md
* Last commit: https://github.com/mdn/content/commit/b6984118ac9482e683a654edfefa4b426ca3c7ca
* Document last modified: 2023-04-07T06:03:34.000Z

</details>

**Comment by randfur** [comment]
Relevant spec section doesn't elaborate as much as MDN attempts to: https://gpuweb.github.io/gpuweb/#dom-gpubindingcommandsmixin-setbindgroup-index-bindgroup-dynamicoffsets-index

**Comment by teoli2003** [comment]
cc/ @chrisdavidmills

**Comment by chrisdavidmills** [comment]
I'm not 100% sure what the correct answer is here. When writing the docs, I found the interplay between `createbindGroup()`/`GPUBindGroup()`, `createbindGroupLayout()`/`GPUBindGroupLayout`, and `setBindGroup()` really difficult to make sense of.

The first two methods both take entries in their descriptors, which reference `@binding`s in the WGSL code, so they are basically saying these are the resources to bind to these specific bindings in the shaders.

In addition, pipeline layouts reference the `@group`s in the shaders; see https://gpuweb.github.io/gpuweb/#pipeline-layout-creation.

Basically the question is "when we say _The index to set the bind group at_, at https://developer.mozilla.org/en-US/docs/Web/API/GPUComputePassEncoder/setBindGroup#parameters, what actually is this index, and what does it correspond to?

@toji , can you help us out here?

**Comment by chrisdavidmills** [comment]
I also wonder whether we should try to create a separate article that very clearly explains all of these objects, how they work together, what parts of the shader code they reference, etc.?

**Comment by toji** [comment]
Sorry I missed this on the initial review.

With `setBindGroup()` the index given is indeed the _`@group`_ index, not the _`@binding`_ index. The `@binding` index of a resource is determined by the `binding` value of each entry in the `createBindGroupLayout()` and `createBindGroup()` calls.

**Comment by chrisdavidmills** [comment]
> Sorry I missed this on the initial review.
> 
> With `setBindGroup()` the index given is indeed the _`@group`_ index, not the _`@binding`_ index. The `@binding` index of a resource is determined by the `binding` value of each entry in the `createBindGroupLayout()` and `createBindGroup()` calls.
> 
> This would need to be updated for the `setBindGroup()` methods on `GPUComputePassEncoder`, `GPURenderPassEncoder`, and `GPURenderBundleEncoder`.

No worries, thanks for the help! I'll raise a quick PR to fix this.
